server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java

```diff
@@ -187,10 +198,15 @@ public ClientQuerySegmentWalker(
   @Override
   public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> query, Iterable<SegmentDescriptor> specs)
   {
-    // Inlining isn't done for segments-based queries.
+    // Inlining isn't done for segments-based queries, but we still globalify the table datasources if possible
+    final Query<T> freeTradeQuery = query.withDataSource(globalizeIfPossible(query.getDataSource()));
 
     if (canRunQueryUsingClusterWalker(query)) {
-      return decorateClusterRunner(query, clusterClient.getQueryRunnerForSegments(query, specs));
+      return new QuerySwappingQueryRunner<>(
+          decorateClusterRunner(freeTradeQuery, clusterClient.getQueryRunnerForSegments(query, specs)),
```

Comment:
super nit: even though `query` is not used in `CachingClusteredClient.getQueryRunnerForSegments()`, it could be better to use `freeTradeQuery` instead of `query` for future proof for when someone changes the logic of `CachingClusteredClient.getQueryRunnerForSegments()`.

Reply:
oops, good catch, fixed :+1: